erts/emulator/beam/jit/arm/instr_common.cpp

```diff
@@ -1446,11 +1533,183 @@ void BeamModuleAssembler::emit_i_test_arity(const ArgLabel &Fail,
+            mov_imm(TMP4, NIL);
+            a.ccmp(TMP3, TMP4, imm(NZCV::kNone), imm(arm::CondCode::kEQ));
+            a.b_ne(resolve_beam_label(Fail, disp1MB));
+
+            return;
+        } else if (beam_jit_is_list_of_immediates(literal)) {
+            comment("optimized equality test with %T", literal);
+            mov_var(ARG1, x);
+            mov_arg(ARG2, Y);
+            fragment_call(ga->get_is_eq_exact_list_shared());
+            a.b_ne(resolve_beam_label(Fail, disp1MB));
+
+            return;
+        } else if (beam_jit_is_shallow_boxed(literal)) {
+            comment("optimized equality test with %T", literal);
+            mov_var(ARG1, x);
+            mov_arg(ARG2, Y);
+            fragment_call(ga->get_is_eq_exact_shallow_boxed_shared());
+            a.b_ne(resolve_beam_label(Fail, disp1MB));
+
+            return;
+        } else if (is_bitstring(literal) && bitstring_size(literal) == 0) {
+            Label not_sub_bits = a.newLabel();
+
+            comment("simplified non-equality test with empty binary");
+            emit_is_boxed(resolve_beam_label(Fail, dispUnknown), X, x.reg);
+            emit_untag_ptr(ARG1, x.reg);
+
+            ERTS_CT_ASSERT_FIELD_PAIR(ErlHeapBits, thing_word, size);
+            a.ldp(TMP1, TMP2, arm::Mem(ARG1));
+
+            a.cmp(TMP1, imm(HEADER_SUB_BITS));
+            a.b_ne(not_sub_bits);
+            {
+                ERTS_CT_ASSERT_FIELD_PAIR(ErlSubBits, start, end);
+                a.ldp(TMP2, TMP3, arm::Mem(ARG1, offsetof(ErlSubBits, start)));
+                a.sub(TMP2, TMP3, TMP2);
+            }
+            a.bind(not_sub_bits);
+
+            if (masked_types<BeamTypeId::MaybeBoxed>(X) ==
+                BeamTypeId::Bitstring) {
+                comment("skipped header test since we know it's a bitstring "
+                        "when boxed");
+                a.cbnz(TMP2, resolve_beam_label(Fail, disp1MB));
+            } else {
+                const auto mask = _BITSTRING_TAG_MASK & ~_TAG_PRIMARY_MASK;
+                ERTS_CT_ASSERT(TAG_PRIMARY_HEADER == 0);
+                ERTS_CT_ASSERT(_TAG_HEADER_HEAP_BITS ==
+                               (_TAG_HEADER_HEAP_BITS & mask));
+                a.and_(TMP1, TMP1, imm(mask));
+                a.cmp(TMP1, imm(_TAG_HEADER_HEAP_BITS));
+                a.ccmp(TMP2, imm(0), imm(NZCV::kNone), imm(arm::CondCode::kEQ));
+                a.b_ne(resolve_beam_label(Fail, disp1MB));
+            }
+
+            return;
+        } else if (is_map(literal) && erts_map_size(literal) == 0) {
+            comment("optimized equality test with empty map", literal);
+            emit_is_boxed(resolve_beam_label(Fail, dispUnknown), X, x.reg);
+            emit_untag_ptr(ARG1, x.reg);
+            a.ldp(TMP1, TMP2, arm::Mem(ARG1));
+            cmp(TMP1, MAP_HEADER_FLATMAP);
+            a.ccmp(TMP2, imm(0), imm(NZCV::kNone), imm(arm::CondCode::kEQ));
+            a.b_ne(resolve_beam_label(Fail, disp1MB));
```

Comment:
We could reduce this to a `LDR`+`CMP` pair if we put the flatmap size in the header word instead of storing it separately. There doesn't seem to be anything preventing us from doing so.

Reply:
Yes, seems reasonable. I think that will better as a separate pull request.

Reply:
Thinking a little bit more about it, some operations such as `map_size/1` would become more complicated because we would have to test which type of map it is. Not really sure that the saving of one word is worth it.

Reply:
I'm not so sure. We already need to care about the map type for all operations other than `map_size/1`, and the extra handling for that doesn't strike me as too difficult. Saving a word off a very common datatype might be worthwhile.